docs/pallas/tpu/distributed.md

```diff
@@ -0,0 +1,1453 @@
+
+By default, if you do not specify a barrier semaphore, Pallas will automatically insert a barrier semaphore at the beginning of your program. However, it can be more efficient to write your own. Barrier semaphores are similar to regular semaphores in that they are counters that can be incremented via `semaphore_signal` and can be decremented via `semaphore_wait`. They are created by calling `get_barrier_semaphore()` within a kernel. Typically, we use barriers once at the beginning of a kernel to synchronize with all devices we are communicating with.
+
+```python
+from jax.experimental.pallas import tpu as pltpu
+
+def example_kernel(...):
+  # Use barrier semaphores at the beginning of a kernel.
+  # is_start_of_kernel = ...
+  # right_neighbor = ...
+  # ...
+  @pl.when(is_start_of_kernel)
+  def _():
+    barrier_sem = pltpu.get_barrier_semaphore()
+    # Increment the semaphore of your right neighbor.
+    pltpu.semaphore_signal(
+          barrier_sem,
+          device_id=right_neighbor,
+          device_id_type=pltpu.DeviceIdType.LOGICAL,
+    )
+    # Wait until your left neighbor has incremented your semaphore
+    pltpu.semaphore_wait(barrier_sem, 1)
+  # ...
+```
+
+When using barrier semaphores, the `collective_id` compiler parameter must be passed to `pallas_call` to specify which barrier semaphore is being used. A TPU has a small, fixed number of barrier semaphores available (typically on the order of 20-30) and therefore they should be used sparingly. In order to ensure correctness, only kernels that share the same communication pattern should use the same `collective_id`. For example, if two kernels synchronize only with neighbors on the same mesh axis, they are allowed to share the same `collective_id`. However, if two kernels synchronize along different axes, they must have different `collective_id`s. Failure to do so may result in race conditions that are difficult to debug.
+
+```python
+kernel = pl.pallas_call(
+      example_kernel,
+      ...,
+      compiler_params=dict(mosaic=dict(collective_id=0)),
+)
+```
+
++++ {"id": "zy20AxN5TSLA"}
+
+### Double-buffering
+
+In order to avoid reading from a local Ref that is also being written into by another device and creating a race condition, a useful technique is the "double-buffered" strategy where we allocate a two `Ref`s for each destination value. On each iteration, one `Ref` will be designated as a "working" slot, and the other will be designated as a "receiving" slot. The device is free to use the working slot for computation, but will only copy data into its neighbor's receiving slot. The working and receiving slots alternate every iteration, so that once a copy is finished, the old receiving slot becomes the new working slot, and vice versa. Using this scheme properly, data is never read from and written to the same buffer.
```

Comment:
```suggestion
In order to avoid reading from a local `Ref` that is also being written into by another device and creating a race condition, a useful technique is the "double-buffered" strategy where we allocate a two `Ref`s for each destination value. On each iteration, one `Ref` will be designated as a "working" slot, and the other will be designated as a "receiving" slot. The device is free to use the working slot for computation, but will only copy data into its neighbor's receiving slot. The working and receiving slots alternate every iteration, so that once a copy is finished, the old receiving slot becomes the new working slot, and vice versa. Using this scheme properly, data is never read from and written to the same buffer.
```

Reply:
Done.